optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,397 @@
+    two = xi * d
+    three = np.zeros(len(es[0]))
+    for i in range(n - 1):
+        three += etas[i] * es[i]
+    three *= D
+    return one + two + three
+
+
+def _normalized_x1_to_x2(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:
+    v_12 = x2 - x1
+    m_12 = np.linalg.norm(v_12, ord=2)
+    e_12 = v_12 / np.clip(m_12, 1e-10, None)
+    return e_12
+
+
+def _distance_from_x_to_PSL(x1: np.ndarray, x2: np.ndarray, x3: np.ndarray) -> np.ndarray:
+    e_12 = _normalized_x1_to_x2(x1, x2)
+    v_13 = x3 - x1
+    v_12_3 = v_13 - np.dot(v_13, e_12) * e_12
+    m_12_3 = np.linalg.norm(v_12_3, ord=2)
+    return m_12_3
+
+
+def _orthonormal_basis_vector_to_PSL(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:
+    n = len(x1)
+    e_12 = _normalized_x1_to_x2(x1, x2)
+    basis_matrix = np.identity(n)
+    if np.count_nonzero(e_12) != 0:
+        v_01 = x1 - np.zeros(len(x1))
+        basis_matrix[0] = v_01 - np.dot(v_01, e_12) * e_12
+    basis_matrix_t = basis_matrix.T
+    Q, _ = np.linalg.qr(basis_matrix_t)
+    return Q.T
+
+
+def _undxm(
+    xs: np.ndarray, rng: np.random.RandomState, sigma_xi: float, sigma_eta: float
+) -> np.ndarray:
+    # https://www.jstage.jst.go.jp/article/sicetr1965/36/10/36_10_875/_pdf
+
+    x_mp2, xs = xs[-1], xs[:-1]  # (1), (3)
+    m = len(xs) - 1
+    dim = len(x_mp2)
+    p = np.sum(xs, axis=0) / (m + 1)  # (2)
+    ds = [x - p for x in xs]  # (2)
+    n = normal(rng, ds[:-1])  # Normal to the plane that contains d_i(1,..,m) (4)
+    d_mp2 = x_mp2 - p  # (4)
+    D = np.dot(d_mp2, n) / np.linalg.norm(n)  # (4)
+    es = _orthonormal_basis_vector_from_ds(
+        ds[:-1]
+    )  # orthonormal basis of the subspace orthogonal to d_i(1,..,m) (5)
+
+    one = p
+
+    ws = rng.normal(0, sigma_xi ** 2, size=m)
+    two = np.zeros(dim)
+
+    for i in range(m):
+        two += ws[i] * ds[i]
+
+    three = np.zeros(dim)
+    for i in range(dim - m):
+        vs = rng.normal(0, sigma_eta ** 2, size=dim)
+        three += vs - np.dot(vs, es[i]) * es[i]
+    three *= D
+    return one + two + three
+
+
+def normal(rng: np.random.RandomState, ds: List[np.ndarray]) -> np.ndarray:
+    d = rng.normal(0, 1, size=ds[0].shape[0])
+    ds.append(d)
+    X = np.stack(ds)
+    Q, _ = np.linalg.qr(X.T)
+    return Q.T[-1]
+
+
+def _orthonormal_basis_vector_from_ds(ds: List[np.ndarray]) -> np.ndarray:
+    X = np.stack(ds)
+    Q, _ = np.linalg.qr(X.T)
+    return Q.T[-1]
+
+
+def _spx(xs: np.ndarray, rng: np.random.RandomState, epsilon: float) -> np.ndarray:
+    # https://www.smapip.is.tohoku.ac.jp/~smapip/2003/tutorial/textbook/hajime-kita.pdf
+    n = xs.shape[0] - 1
+    G = xs.sum(axis=0) / xs.shape[0]
+    rs = [np.power(rng.uniform(0, 1), 1 / (k + 1)) for k in range(n)]
+    xks = [G + epsilon * (pk - G) for pk in xs]
+    ck = 0
+    for k in range(1, n + 1):
+        ck = rs[k - 1] * (xks[k - 1] - xks[k] + ck)
+    c = xks[-1] + ck
+    return c
+
+
+def _check_in_constraints(params: Dict, search_space: Dict) -> bool:
```

Comment:
This is a nit given it's a private function, but what do you think about the following change? 
```suggestion
def _is_contained(params: Dict, search_space: Dict) -> bool:
```
"check" usually in Optuna communicates that it checks for something and raises if violated, while "is" communicates that the return is a bool.